### PR TITLE
fix returning error from local invocation

### LINKED ISSUE #2612: Serverless local invoke failing to capture a purposeful failure
# This is a Bug Report

## Description

For bug reports:
* What went wrong?
I locally invoked a function to purposely fail and serverless threw an error

* What did you expect should have happened?
I expected serverless handle the purposeful failure the same way lambda might
```
serverless-local-invoke-failure master % serverless invoke -f hello
{
    "errorMessage": "fail"
}
```

* What was the config you used?
Code can be found here: https://github.com/jtslear/serverless-local-invoke-failure

* What stacktrace or error message from your provider did you see?
```
serverless-local-invoke-failure master % SLS_DEBUG=true serverless invoke local -f hello

  Error --------------------------------------------------

     fail

     For debugging logs, run again after setting SLS_DEBUG env var.

  Stack Trace --------------------------------------------

Error: fail
    at module.exports.logError.words.forEach (/usr/local/lib/node_modules/serverless/lib/classes/Error.js:82:11)
    at /usr/local/lib/node_modules/serverless/bin/serverless:20:15
    at tryCatcher (/usr/local/lib/node_modules/serverless/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/usr/local/lib/node_modules/serverless/node_modules/bluebird/js/release/promise.js:510:31)
    at Promise._settlePromise (/usr/local/lib/node_modules/serverless/node_modules/bluebird/js/release/promise.js:567:18)
    at Promise._settlePromise0 (/usr/local/lib/node_modules/serverless/node_modules/bluebird/js/release/promise.js:612:10)
    at Promise._settlePromises (/usr/local/lib/node_modules/serverless/node_modules/bluebird/js/release/promise.js:687:18)
    at Async._drainQueue (/usr/local/lib/node_modules/serverless/node_modules/bluebird/js/release/async.js:138:16)
    at Async._drainQueues (/usr/local/lib/node_modules/serverless/node_modules/bluebird/js/release/async.js:148:10)
    at Immediate.Async.drainQueues [as _onImmediate] (/usr/local/lib/node_modules/serverless/node_modules/bluebird/js/release/async.js:17:14)
    at processImmediate [as _immediateCallback] (timers.js:383:17)

  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues

     Please report this error. We think it might be a bug.

  Your Environment Information -----------------------------
     OS:                 darwin
     Node Version:       4.3.2
     Serverless Version: 1.1.0
```

## Additional Data

* ***Serverless Framework Version you're using***: 1.1.0
* ***Operating System***: OSX
* ***Stack Trace***: above
* ***Provider Error messages***: above

Unless of course this is expected behavior, feel free to close otherwise.  While building lambda functions I'm purposely failing them for testing, and the `lambci/lambda` docker container handles it like I'd expect, or comfortable with knowing my function failed, and not the serverless local invocation.

### PR DESCRIPTION
**Pull request by mthenw**
## What did you implement:

Closes #2612

This implements AWS Lambda behaviour.

## How can we verify it:

Run `sls invoke local` against function calling back error

## Note

we shouldn't format or color error messages all over the codebase. There should be one place with logs styling. I will create issue for that becuase right now it's like inline css styles.

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config/commands/resources
- [x] Enable ["Allow edits from maintainers"](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) for this PR
- [x] Change ready for review message below

***Is this ready for review?:*** Yes

### CODE REVIEW

**Comment by nikgraf** [comment]
Reviewing now 👍

File: lib/plugins/aws/invokeLocal/index.js [hunk 1]
```diff
@@ -95,7 +96,20 @@ class AwsInvokeLocal {
...
+            errorType: err.constructor.name,
+          };
+        } else {
+          errorResult = {
+            errorMessage: err,
+          };
+        }
+
+        this.serverless.cli.consoleLog(chalk.red(JSON.stringify(errorResult, null, 4)));
+        process.exit(1);
```
Comment (dougmoscrop) [inline lib/plugins/aws/invokeLocal/index.js]: With this PR: https://github.com/serverless/serverless/pull/2836

You should not (have to) process.exit, and if you want to set the exit code, set process.exitCode or (preferably) reject the promise.

Comment (mthenw) [inline lib/plugins/aws/invokeLocal/index.js]: You are absolutely right.

**Comment by QuantumInformation** [comment]
How do you get a useful error from an aync operation, ie this always give `null` as output:
```js
module.exports.test = () => {
  sendMail('t', 't').then((data) => {
    console.log(data)
  }).catch((data) => {
    console.log(data)
  })
}
```

**Comment by pmuens** [comment]
@QuantumInformation how dos the `sendMail` method looks like?

**Comment by QuantumInformation** [comment]
image

**Comment by dougmoscrop** [comment]
I don't know if this is a typo in your example, but in this case your test
function returns *immediately* and the sendMail promise is not returned. So
you have an "orphan" promise.

you want to return sendMail()

**Comment by pmuens** [comment]
Yep. Thanks @dougmoscrop 

you need to do the following:

```javascript
module.exports.sendMail = (event, context, callback) => {
  // NOTE the return here...
  return basicEmail('t', 't').then(() => { ... })
}
```

**Comment by QuantumInformation** [comment]
I'm not really used to handling promises this way, anyway I still have the same issue:

image


works fine in node:

image

**Comment by pmuens** [comment]
@QuantumInformation thanks for the response.

Your function works according to the `console` output as the request is sent to Mailgun and is queued. What issue do you still face?

**Comment by QuantumInformation** [comment]
My apologies the working example is when I call it purely via node, the issue remains when using svs invoke. Skype screen share?

**Comment by pmuens** [comment]
@QuantumInformation `serverless invoke` and it's offline version `serverless invoke local` will execute the function according to the `handler` definition in the `serverless.yml` file.

Could you please share your `serverless.yml` and the corresponding `handler.js` file with its code. Would be even better if you could paste the code here so that it's easier to read.

Thanks!

**Comment by QuantumInformation** [comment]
Sure mate thx:

```js
var nodemailer = require('nodemailer')
var mg = require('nodemailer-mailgun-transport')

module.exports.sendMail = (event, context, callback) => {
 return sendMail('t', 't').then((data) => {
    console.log(data)
  }).catch((data) => {
    console.log(data)
  })
}

var sendMail = (subject, mailBody) => {
  // This is your API key that you retrieve from www.mailgun.com/cp (free up to 10K monthly emails)


  var nodemailerMailgun = nodemailer.createTransport(mg(auth))

  return new Promise(function (resolve, reject) {
    nodemailerMailgun.sendMail({
      from: 'nkatsikanis@gmail.com',
      to: 'nkatsikanis@gmail.com', // An array if you have multiple recipients.
      subject: subject,
      text: mailBody,
    }, function (err, info) {
      if (err) {
        console.log('Error: ' + err)
        reject(err)
      }
      else {
        resolve(info)
        console.log('Response: ' + info)
      }
    })
  })

}
/*
sendMail('t', 't').then((data) => {
  console.log(data)
}).catch((data) => {
  console.log(data)
})
*/

```

```yaml
# Welcome to Serverless!
#
# This file is the main config file for your service.
# It's very minimal at this point and uses default values.
# You can always add more config options for more control.
# We've included some commented out config examples here.
# Just uncomment any of them to get that config option.
#
# For full config options, check the docs:
#    docs.serverless.com
#
# Happy Coding!

service: aws-nodejs # NOTE: update this with your service name

# You can pin your service to only deploy with a specific Serverless version
# Check out our docs for more details
# frameworkVersion: "=X.X.X"

provider:
  name: aws
  runtime: nodejs4.3

# you can overwrite defaults here
#  stage: dev
  region: us-west-2

# you can add statements to the Lambda function's IAM Role here
#  iamRoleStatements:
#    - Effect: "Allow"
#      Action:
#        - "s3:ListBucket"
#      Resource: { "Fn::Join" : ["", ["arn:aws:s3:::", { "Ref" : "ServerlessDeploymentBucket" } ] ]  }
#    - Effect: "Allow"
#      Action:
#        - "s3:PutObject"
#      Resource:
#        Fn::Join:
#          - ""
#          - - "arn:aws:s3:::"
#            - "Ref" : "ServerlessDeploymentBucket"

# you can define service wide environment variables here
#  environment:
#    variable1: value1

# you can add packaging information here
#package:
#  include:
#    - include-me.js
#    - include-me-dir/**
#  exclude:
#    - exclude-me.js
#    - exclude-me-dir/**

functions:
  sendMail:
    handler: emailServices.sendMail

#    The following are a few example events you can configure
#    NOTE: Please make sure to change your handler code to work with those events
#    Check the event documentation for details
#    events:
#      - http:
#          path: users/create
#          method: get
#      - s3: ${env:BUCKET}
#      - schedule: rate(10 minutes)
#      - sns: greeter-topic
#      - stream: arn:aws:dynamodb:region:XXXXXX:table/foo/stream/1970-01-01T00:00:00.000
#      - alexaSkill
#      - iot:
#          sql: "SELECT * FROM 'some_topic'"

#    Define function environment variables here
#    environment:
#      variable2: value2

# you can add CloudFormation resource templates here
#resources:
#  Resources:
#    NewResource:
#      Type: AWS::S3::Bucket
#      Properties:
#        BucketName: my-new-bucket
#  Outputs:
#     NewOutput:
#       Description: "Description for the output"
#       Value: "Some output value"
```
image

**Comment by pmuens** [comment]
OK. Your seems to be fine. So you're trying to invoke that locally, right? With `serverless invoke local`?

`node emailServices.js` works fine whereas `serverless invoke local sendMail` fails?

/ccing @eahefnawy here as he's implemented local invocation...

**Comment by QuantumInformation** [comment]
Yes thats right.

**Comment by QuantumInformation** [comment]
Maybe its the promise I used?

**Comment by QuantumInformation** [comment]
That should work locally as I'm using node v6. For cloud I would import bluebird.

**Comment by pmuens** [comment]
@eahefnawy do you have a clue what's going on here?

**Comment by QuantumInformation** [comment]
I removed the promise and still have the same issue:

image

no pressure, but if this can be fixed I can demo this at a AWS meetup in London today perhaps.

**Comment by QuantumInformation** [comment]
Is there any slides I could use in 2hours to give a quick overview of the framework?

**Comment by QuantumInformation** [comment]
I'll present it in March.)

**Comment by pmuens** [comment]
Nice that you present at the London Meetup @QuantumInformation ! 👍 

Let's see. Not the best debugging, but could you maybe put some `console.log` statement in the `sendEmail` function so that we can see where exactly the return is happening?

Do you use correct keys on your machine so that (in theory) you could send E-Mails from your local machine?

**Comment by QuantumInformation** [comment]
Not presenting yet, Having talks with the movers and shakers though. svs would defo be popular.

Would you like to do a skype screen share?

the function never seems to run:
image

**Comment by eahefnawy** [comment]
@QuantumInformation could you confirm that this is an issue with invoke local only or invoking the actual remote lambda as well? What kinda error do you get when invoking the deployed/remote lambda?

**Comment by QuantumInformation** [comment]
The remote works. I get my email:

image
image

Local:
image

**Comment by QuantumInformation** [comment]
I tried changing the name of the function in Yaml, but get this:
image

**Comment by QuantumInformation** [comment]
Sorry I just tried this and am getting a more successful message

➜  time-tracker-frontend git:(master) ✗ serverless invoke local -f sendIt

I forgot to use local. I didn't get the email, but I think there is something else going on the node mailer I can solve.

**Comment by QuantumInformation** [comment]
Is there a way to debug local invoke?
image
btw I've learned my lesson, don't use aliases in github examples)

**Pull request merged**